### ISSUE TITLE
@OAuth2RestOperationsAutowired causes just a BeanCreationException

### ISSUE THREAD

**Issue by ePaul** [zalando-stups/spring-boot-zalando-stups-tokens#3]
Maybe I'm using it wrong, but there isn't much documentation.

From [@jbellmann's comment](https://github.com/zalando-stups/stups-spring-oauth2-support/issues/33#issuecomment-215215806) I did build this:

``` java
@Configuration
public class ProducerConfiguration {

    private static final String PRODUCER_TOKEN_NAME = "producer";

    @Value("${PRODUCER_EVENTS_URI}")
    private String producerEventsURI;

    @OAuth2RestOperationsAutowired(PRODUCER_TOKEN_NAME)
    private RestOperations template;
}
```

In my application-dev.yml I have the following setup:

``` yaml
tokens:
    token-configuration-list:
        - tokenId: producer
          scopes:
              - uid
              - producer.event_stream.read
        - tokenId: nakadi
          scopes: nakadi.event_stream.write


spring:
  oauth2:
    resource:
      tokenInfoUri: https://info.services.auth.zalando.com/oauth2/tokeninfo
```

Starting the app (with a dummy producer URL in an environment variable) gives me just a BeanCreationException:

> Caused by: org.springframework.beans.factory.NoSuchBeanDefinitionException: No qualifying bean of type [org.springframework.web.client.RestOperations] found for dependency: expected at least 1 bean which qualifies as autowire candidate for this dependency. Dependency annotations: {@org.zalando.stups.tokens.annotation.OAuth2RestOperationsAutowired(value=producer)}

This doesn't change whether or not I have also `stups-spring-oauth2-client` in my Maven dependencies.

**Comment by jbellmann** [comment]
Thanks @ePaul,
I will try to reproduce this.

**Comment by jbellmann** [comment]
Your observation is right. It only worked in the test because beans injected in the test-unit itself where injected last. So I have to reimplement that feature.
I will inform you, when I think it will work better.

**Comment by jbellmann** [comment]
@ePaul ,
please try again with 0.9.11.M2.

**Comment by jbellmann** [comment]
And please have look [into this example](https://github.com/zalando-stups/spring-boot-zalando-stups-tokens/tree/oauth2_client_support/examples/autowired-restoperations).

**Comment by ePaul** [comment]
@jbellmann Thanks, my app now starts up without problems. (Not sure if this actually works, I'm not yet this far in the program.)

**Comment by ePaul** [comment]
I can also actually exchange requests, thanks.

**Comment by jbellmann** [comment]
So, before we put this new thing into an official release, do you have any improvements (renaming maybe, annotation to long?) of the current state?

**Comment by ePaul** [comment]
I think it is usable in the current form. (It looks like we can also autowire the `RestTemplate` class directly, which is needed for Riptide.)

A wish (maybe as a separate feature), though I'm not sure if it fits here:

I would like to have some way to access the list of token-configurations, preferably also with other information added inside it. For the example above, I would like to have several producer configurations and several Nakadi ones, and be able to iterate over them, or find them by name. I guess I can put this in a separate configuration section (and just use the name to refer to the token one).

**Comment by jbellmann** [comment]
You are right, RestTemplate is also possible :

```
    @OAuth2RestOperationsAutowired(PRODUCER_TOKEN_NAME)
    private RestTemplate template2;
```

To access the token-configurations just:

```
    @Autowired
    private AccessTokensBeanProperties accessTokensBeanProperties;

... then 

accessTokensBeanProperties.getTokenConfigurationList()

```

Provide an PR for more.

Thanks for contributing.

**Comment by ePaul** [comment]
Hi,

I guess I got another issue. We configured our own HttpComponentsClientHttpRequestFactory as a spring bean (for adding some interceptors for logging and tracing).

But the StupsOAuth2RestTemplate which is being built from the annotation here is using its own request factory (using the one-argument constructor), with the effect that our logging doesn't work.

Could this be made to use the constructor with a request factory, if such a spring-bean is available?
